Add vm export vm

awels opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

**What this PR does / why we need it**:
Add VM source type to VMExport object CRD, including all needed code to support it in the controller and admission webhooks. The export will not start if a running VM is detected (VMI exists) and will be stopped if a VMI is detected associated with the VM after the export started.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:
Addressed the style comment noted in the VMSnapshot PR.

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
VirtualMachineExport now supports VM export source type.
```

awels commented:
/test pull-kubevirt-goveralls

awels commented:
/test pull-kubevirt-e2e-k8s-1.22-sig-compute-migrations

mhenriks commented on tests/storage/export.go:
```diff
@@ -1211,14 +1274,54 @@ var _ = SIGDescribe("Export", func() {
+		export := createVMExportObject(vm.Name, vm.Namespace, token)
+		Expect(export).ToNot(BeNil())
+		waitForExportPhase(export, exportv1.Skipped)
 	})
```
Would like an additional test case for export running, vm started, export cancelled

mhenriks commented on pkg/virt-controller/watch/export/export.go:
```diff
@@ -111,6 +112,8 @@ const (
 
 	exporterPodFailedOrCompletedEvent = "ExporterPodFailedOrCompleted"
 	exporterPodCreatedEvent           = "ExporterPodCreated"
+	vmStartedEvent                    = "ExportedVMStarted"
```
hmm, maybe generic `ExportPaused` with more detail in message

awels replied:
Done, if I understood you correctly.

mhenriks commented on pkg/virt-controller/watch/export/export.go:
```diff
@@ -961,6 +1109,22 @@ func (ctrl *VMExportController) updateVMExporVMSnapshotStatus(vmExport *exportv1
 	return nil
 }
 
+func (ctrl *VMExportController) updateVMExporVMStatus(vmExport *exportv1.VirtualMachineExport, pvcs []*corev1.PersistentVolumeClaim, exporterPod *corev1.Pod, service *corev1.Service, pvcInUse bool) error {
```
`updateVMExpor` t` VMStatus`

awels replied:
Fixed

mhenriks commented on pkg/virt-controller/watch/export/export.go:
```diff
@@ -507,11 +570,94 @@ func (ctrl *VMExportController) handleIsVmSnapshot(vmExport *exportv1.VirtualMac
+		if exists {
+			pvcs = append(pvcs, pvc)
+		}
+		name = ""
```
Why not just declare this variable in the loop?

awels replied:
Don't know changed.

mhenriks commented on pkg/virt-controller/watch/export/export.go:
```diff
@@ -445,23 +506,29 @@ func (ctrl *VMExportController) handleIsSourcePvc(vmExport *exportv1.VirtualMach
 		}
 	} else {
-		if err := ctrl.handlePodSucceededOrFailed(vmExport, pod); err != nil {
+		inUse, err = ctrl.isPVCInUse(vmExport, pvc)
```
I'm not sure we want to do this check.  As-is it will be triggered randomly.  We are not watching all pods and triggering reconcile when an exported PVC is referenced in a pod.  And I don't think we care to do that.

awels replied:
Well my main idea behind this one is that we are watching VMIs, and if someone has exported a PVC (instead of a VM) that at least when VMs are started this will get triggered.

mhenriks replied:
> Well my main idea behind this one is that we are watching VMIs, and if someone has exported a PVC (instead of a VM) that at least when VMs are started this will get triggered.

I don't think that will work as-is.  The VMI watch queues exports that reference a VM.  It does not queue exports with PVC source.  You would have to go through all the PVCs in the VMI and see if there is an export.

Also if you want to support this case, add a functional test.

mhenriks commented on pkg/virt-controller/watch/export/export.go:
```diff
@@ -507,11 +570,94 @@ func (ctrl *VMExportController) handleIsVmSnapshot(vmExport *exportv1.VirtualMac
+	var pvcs []*corev1.PersistentVolumeClaim
+
+	inUse := false
+	vmi, exists, err := ctrl.getVmi(vmExport.Namespace, vmExport.Spec.Source.Name)
```
Not watching/triggering reconcile on VMIs.  So let's say an export is started while a VM is shutting down.  Will not be notified when the VMI is gone

mhenriks reviewed (commented):
Great start!

awels commented:
/test pull-kubevirt-e2e-k8s-1.23-sig-compute-migrations

awels commented:
/test pull-kubevirt-code-lint

mhenriks commented on pkg/storage/export/export/export.go:
```diff
@@ -453,8 +453,42 @@ func (ctrl *VMExportController) handleVMI(obj interface{}) {
+		if volume.DataVolume != nil {
+			if pvc := ctrl.getPVCsFromName(vmi.Namespace, volume.DataVolume.Name); pvc != nil {
+				pvcs = append(pvcs, pvc)
+			}
```
What about memDump?  Also, could use `PVCNameFromVirtVolume` helper

awels replied:
mem dumps are associated with VMs, not VMIs which is why I didn't include them in this check. I didn't know PVCNameFromVirtVolume existed, let me see if that can work.

mhenriks commented on pkg/storage/export/export/export.go:
```diff
@@ -417,52 +530,82 @@ func (ctrl *VMExportController) updateVMExport(vmExport *exportv1.VirtualMachine
 		}
 	} else {
-		if err := ctrl.handlePodSucceededOrFailed(vmExport, pod); err != nil {
+		inUse, err = ctrl.isPVCInUse(vmExport, pvc)
```
`isPVCInUse` is called first thing in both branches of this if statement.  So maybe do it before?  Also, if not `isPopulated` maybe infer inUse and not do a bunch of stuff here.  Maybe could clean up some of the conditional branches

awels replied:
Sure let me see if I can clean that up.

mhenriks commented on pkg/storage/export/export/export.go:
```diff
@@ -508,11 +647,93 @@ func (ctrl *VMExportController) handleIsVmSnapshot(vmExport *exportv1.VirtualMac
+	}
+	if !exists {
+		if !inUse {
+			pod, err = ctrl.createPodAndSecret(vmExport, pvcs)
```
I think it is problematic to combine these two operations.  What if creating the pod succeeds but creating the secret fails?  Next time through the reconcile you will see the pod exists and will not attempt to recreate the secret.  This applies to all sources, not just VM, of course

mhenriks commented on pkg/storage/export/export/export.go:
```diff
@@ -508,11 +647,93 @@ func (ctrl *VMExportController) handleIsVmSnapshot(vmExport *exportv1.VirtualMac
+	var pvcs []*corev1.PersistentVolumeClaim
+
+	inUse := false
+	vmi, exists, err := ctrl.getVmi(vmExport.Namespace, vmExport.Spec.Source.Name)
```
Maybe get the vm as well to make sure not in "provisioning" state?

mhenriks commented on pkg/storage/export/export/export.go:
```diff
@@ -416,103 +524,194 @@ func (ctrl *VMExportController) updateVMExport(vmExport *exportv1.VirtualMachine
+		return nil, nil
+	}
+	for _, volume := range vm.Spec.Template.Spec.Volumes {
+		pvc, exists, err := ctrl.getPvc(vmNamespace, storagetypes.PVCNameFromVirtVolume(&volume))
```
`PVCNameFromVirtVolume` may return ""

awels replied:
Right which then will return !exists from ctrl.getPvc due to the blank name.

mhenriks commented on pkg/storage/export/export/export.go:
```diff
@@ -416,103 +524,194 @@ func (ctrl *VMExportController) updateVMExport(vmExport *exportv1.VirtualMachine
 		}
 	}
-	return 0, ctrl.updateVMExporVMSnapshotStatus(vmExport, pvcs, pod, service, false)
+	return pvcs, nil
```
I still don't see where/how "populating" DVs are detected/handled with VM source

mhenriks reviewed (commented):
Some things to consider

mhenriks commented on pkg/storage/export/export/vm-source.go:
```diff
@@ -0,0 +1,244 @@
+	if err != nil {
+		return &sourceVolumes{}, err
+	}
+	if pvcs != nil && !allPopulated {
```
awkward to check if a slice is != nil.  Checking length is the idiomatic way

awels replied:
Done

mhenriks commented on tests/storage/export.go:
```diff
@@ -1194,14 +1277,116 @@ var _ = SIGDescribe("Export", func() {
+		Expect(export).ToNot(BeNil())
+		waitForExportPhase(export, exportv1.Skipped)
 	})
 })
```
I think we need a test for VM with no (initial?) persistent volumes.  How should that be handled?  Pending?  Skipped?

awels replied:
Should be skipped based on the unit test. I can add a functional test for it.

mhenriks reviewed (commented):
Getting there

mhenriks commented:
/retest-required

mhenriks commented on pkg/storage/export/export/vm-source.go:
```diff
@@ -0,0 +1,245 @@
+	return pvcs, allPopulated, nil
+}
+
+func (ctrl *VMExportController) updateVMExportVMStatus(vmExport *exportv1.VirtualMachineExport, exporterPod *corev1.Pod, service *corev1.Service, sourceVolumes *sourceVolumes) (time.Duration, error) {
```
I believe this function should return a requeue duration when pvcs are populating

awels replied:
You are correct, fixing that.

awels commented:
/test pull-kubevirt-goveralls

mhenriks commented on pkg/storage/export/export/vm-source.go:
```diff
@@ -146,9 +146,9 @@ func (ctrl *VMExportController) isSourceAvailableVM(vmExport *exportv1.VirtualMa
 	}
 	if exists {
-		return false, fmt.Sprintf("Virtual Machine %s/%s is running", vmi.Namespace, vmi.Name), nil
+		return exists, fmt.Sprintf("Virtual Machine %s/%s is running", vmi.Namespace, vmi.Name), nil
```
why return true here?

awels replied:
I should probably rename that function to isSourceInUse since that is what it is doing.

mhenriks commented:
/lgtm
/approve

mhenriks commented:
/retest-required

mhenriks commented:
/hold

awels commented:
/unhold

awels commented:
/test pull-kubevirt-e2e-k8s-1.23-sig-storage

awels commented:
/test pull-kubevirt-e2e-k8s-1.24-sig-compute
/test pull-kubevirt-e2e-k8s-1.23-sig-storage

awels commented:
1 good run
/retest

awels commented:
/test pull-kubevirt-e2e-k8s-1.22-sig-storage
/test pull-kubevirt-e2e-k8s-1.23-sig-storage 
/test pull-kubevirt-e2e-k8s-1.24-sig-storage

awels commented:
2 good runs
/test pull-kubevirt-e2e-k8s-1.22-sig-storage
/test pull-kubevirt-e2e-k8s-1.23-sig-storage
/test pull-kubevirt-e2e-k8s-1.24-sig-storage

awels commented:
3 good runs
/test pull-kubevirt-e2e-k8s-1.22-sig-storage
/test pull-kubevirt-e2e-k8s-1.23-sig-storage
/test pull-kubevirt-e2e-k8s-1.24-sig-storage

awels commented:
4 good runs
/test pull-kubevirt-e2e-k8s-1.22-sig-storage
/test pull-kubevirt-e2e-k8s-1.23-sig-storage
/test pull-kubevirt-e2e-k8s-1.24-sig-storage

awels commented:
5 good runs
/test pull-kubevirt-e2e-k8s-1.22-sig-storage
/test pull-kubevirt-e2e-k8s-1.23-sig-storage
/test pull-kubevirt-e2e-k8s-1.24-sig-storage

awels commented:
1 failure on the test due to the VM getting updated between the get and the update. Added an eventually to solve that. Not related to the issue I fixed

awels commented:
/retest
